toolchain/base/value_store.h

```diff
@@ -186,16 +203,51 @@ class ValueStore<StringId> : public Yaml::Printable<ValueStore<StringId>> {
   llvm::SmallVector<llvm::StringRef> values_;
 };
 
+// A thin wrapper around a `ValueStore<StringId>` that provides a different IdT,
+// while using a unified storage for values. This avoids potentially
+// duplicative string hash maps, which are expensive.
+template <typename IdT>
+class StringStoreWrapper : public Printable<StringStoreWrapper<IdT>> {
+ public:
+  explicit StringStoreWrapper(ValueStore<StringId>* values) : values_(values) {}
+
+  auto Add(llvm::StringRef value) -> IdT {
+    return IdT(values_->Add(value).index);
+  }
+
+  auto Get(IdT id) const -> llvm::StringRef {
+    return values_->Get(StringId(id.index));
+  }
+
+  auto Print(llvm::raw_ostream& out) const -> void { out << *values_; }
+
+ private:
+  ValueStore<StringId>* values_;
+};
+
 // Stores that will be used across compiler phases for a given compilation unit.
 // This is provided mainly so that they don't need to be passed separately.
 class SharedValueStores : public Yaml::Printable<SharedValueStores> {
```

Comment:
We should make this class non-copyable, because the implicitly copy constructor is no longer correct (the copies of the wrappers would end up pointing to the string store in the original).

Alternatively, we could create the wrappers on the fly like I did in #3341, rather than storing an instance. That's also probably slightly more efficient, because it avoids loading the pointer to the string value store from the stored wrapper, although I doubt that would make any real difference in practice.

Reply:
> We should make this class non-copyable, because the implicitly copy constructor is no longer correct (the copies of the wrappers would end up pointing to the string store in the original).

Making non-copyable. I think incidental copies would only lead to problems anyways (e.g., if you assume you can just copy this and a TokenizedBuffer using it, _that_ fails too).

> Alternatively, we could create the wrappers on the fly like I did in #3341, rather than storing an instance. That's also probably slightly more efficient, because it avoids loading the pointer to the string value store from the stored wrapper, although I doubt that would make any real difference in practice.

I'll leave it to you if you want to swap this, but I feel slightly faster to reason about this behavior unless we're going to do the on-the-fly creation as a more common semantic.